### PR TITLE
bug 1630254 updating downgrade for 3.11

### PR DESCRIPTION
**Pull request by kalexand-rh**
https://bugzilla.redhat.com/show_bug.cgi?id=1630254

### CODE REVIEW

**Comment by kalexand-rh** [comment]
@jiajliu, will you PTAL? I'm still waiting for an update on the etcd restore steps, but I'll update the PR as soon as I have it.

**Comment by spahl** [comment]
@joelegasse could you PTAL?

File: upgrading/downgrade.adoc [hunk 1]
```diff
@@ -13,45 +13,40 @@
...
+*_scheduler.json_* file, and the etcd data directory exist on your masters:
 +
 ----
 /etc/origin/master/master-config.yaml.<timestamp>
 /etc/origin/master/scheduler.json
+/var/lib/etcd/openshift-backup-xxxx
 ----
++ The Ansible playbook that you run during the
+xref:../upgrading/index.adoc#install-config-upgrading-index[upgrade process]
+creates these files.
```
Comment (jiajliu) [inline upgrading/downgrade.adoc]: According to "Preparing for an upgrade" in [in-place upgrades], above three files were backup manually. So here should be corresponding with there, and if need a reminder, just notice user that these files would be backup by ansible playbook during upgrade if user forgot to backup them.

File: upgrading/downgrade.adoc [hunk 2]
```diff
@@ -63,24 +58,23 @@ On node and master hosts:
...
 ----
 # systemctl stop atomic-openshift-node
 ----
 
 [[downgrade-removing-rpms]]
-== Removing RPMs and Static Pods
+== Removing RPMs and static pods
 
 The **-excluder* packages add entries to the exclude directive in the host’s
 *_/etc/yum.conf_* file when installed.
```
Comment (jiajliu) [inline upgrading/downgrade.adoc]: This sentence seem not related with following context. Please confirm.

File: upgrading/downgrade.adoc [hunk 3]
```diff
@@ -89,14 +83,7 @@ The **-excluder* packages add entries to the exclude directive in the host’s
...
-# yum remove atomic-openshift \
-    atomic-openshift-clients \
-    atomic-openshift-node \
-    atomic-openshift-master \
-    atomic-openshift-sdn-ovs \
-    atomic-openshift-excluder \
-    atomic-openshift-docker-excluder \
-    atomic-openshift-hyperkube
+# yum remove atomic-openshift-node \
 ----
```
Comment (jiajliu) [inline upgrading/downgrade.adoc]: All high version pkgs need to be removed:
atomic-openshift-excluder-3.11.16-1.git.0.b48b8f8.el7.noarch
atomic-openshift-hyperkube-3.11.16-1.git.0.b48b8f8.el7.x86_64
atomic-openshift-node-3.11.16-1.git.0.b48b8f8.el7.x86_64
atomic-openshift-3.11.16-1.git.0.b48b8f8.el7.x86_64
atomic-openshift-docker-excluder-3.11.16-1.git.0.b48b8f8.el7.noarch
atomic-openshift-clients-3.11.16-1.git.0.b48b8f8.el7.x86_64

File: upgrading/downgrade.adoc [hunk 4]
```diff
@@ -118,22 +105,17 @@ remove the following packages:
...
 +
 ----
 # subscription-manager repos \
-    --disable=rhel-7-server-ose-3.10-rpms \
-    --enable=rhel-7-server-ose-3.9-rpms
+    --disable=rhel-7-server-ose-3.11-rpms \
+    --enable=rhel-7-server-ose-3.10-rpms
 ----
 
 . On each master, install the following packages:
```
Comment (jiajliu) [inline upgrading/downgrade.adoc]: Following list need to be updated.
atomic-openshift-node-3.10.51-1.git.0.90663d3.el7.x86_64
atomic-openshift-docker-excluder-3.10.51-1.git.0.90663d3.el7.noarch
atomic-openshift-excluder-3.10.51-1.git.0.90663d3.el7.noarch
atomic-openshift-clients-3.10.51-1.git.0.90663d3.el7.x86_64
atomic-openshift-hyperkube-3.10.51-1.git.0.90663d3.el7.x86_64
atomic-openshift-3.10.51-1.git.0.90663d3.el7.x86_64

Following sections on each node need to be updated too.

File: upgrading/downgrade.adoc [hunk 5]
```diff
@@ -176,9 +158,11 @@ include::day_two_guide/topics/proc_adding-etcd-after-restoring.adoc[leveloffset=
 include::day_two_guide/topics/proc_bringing-openshift-online.adoc[leveloffset=+1]
```
Comment (jiajliu) [inline upgrading/downgrade.adoc]: This section need to be updated and confirmed with dev due to many config files are not avaiablable now. I'm not sure if it's enough.

Comment (jiajliu) [reply upgrading/downgrade.adoc]: Etcd service is not running as systemd service in v3.10. So when downgrade to v3.10, we can not use original way to unmask/stop/start way to control etcd. Here need to be re-wroten.

Comment (jiajliu) [reply upgrading/downgrade.adoc]: > # rpm -q openvswitch
Openvswitch service is not run as a systemd service. No here no need to verify it.

File: upgrading/downgrade.adoc (hunk 1 above)
Comment (kalexand-rh) [reply upgrading/downgrade.adoc]: Would it be better to point out in the upgrading topic that these files are backed up by the playbook?

File: upgrading/downgrade.adoc (hunk 2 above)
Comment (kalexand-rh) [reply upgrading/downgrade.adoc]: You're right. Thank you!

File: upgrading/downgrade.adoc (hunk 3 above)
Comment (kalexand-rh) [reply upgrading/downgrade.adoc]: Is it ok to only list the package names, not the specific versions?

File: upgrading/downgrade.adoc (hunk 5 above)
Comment (kalexand-rh) [reply upgrading/downgrade.adoc]: Yes, I have an urgent request out to the devs and dev leadership to provide the updated etcd restore information, and I will update this PR as soon as I have new information.

**Comment by kalexand-rh** [comment]
@jiajliu, I made some updates. Will you PTAL? As soon as I get more information about the static pod restore, I will update the docs.

File: upgrading/downgrade.adoc (hunk 3 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: > Is it ok to only list the package names, not the specific versions?

yes, sure. I just list a example for u. You can just list the name.

File: upgrading/downgrade.adoc (hunk 1 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: > Would it be better to point out in the upgrading topic that these files are backed up by the playbook?

Yes. My suggestion is that not only saying our playbook will take a backup of these files, but also ensure user to backup them manually in case of any sudden(refer to https://bugzilla.redhat.com/show_bug.cgi?id=1613077#c3). So when it came to downgrade, they can use their manual backup as first choice, and if not, automatic backup is a last guarantee.

File: upgrading/downgrade.adoc [hunk 6]
```diff
@@ -63,24 +58,23 @@ On node and master hosts:
 /etc/origin/node/node-config.yaml
```
[the remark this answers is not in the document]
Comment (jiajliu) [reply upgrading/downgrade.adoc]: LGTM now.

File: upgrading/downgrade.adoc (hunk 5 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: > > rpm -q openvswitch
> 
> Openvswitch service is not run as a systemd service. No here no need to verify it.

LGTM now

File: upgrading/downgrade.adoc (hunk 4 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: when re-install old version openshift, master and node should install the same pkgs, so suggest to combine "On each master, install the following packages:" and "On each node, install the following packages:"

**Comment by kalexand-rh** [comment]
@jiajliu, will you PTAL at these updates? 

I made the static pod updates on this PR: https://github.com/openshift/openshift-docs/pull/12322

File: upgrading/downgrade.adoc (hunk 1 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: LGTM now

File: upgrading/downgrade.adoc [hunk 7]
```diff
@@ -118,55 +98,38 @@ remove the following packages:
...
-
-. On each node, install the following packages:
-+
-----
-# yum install atomic-openshift \
-    atomic-openshift-node \
+    atomic-openshift-clients \
+    atomic-openshift-hyperkube \
+    atomic-openshift-master \
     openvswitch \
```
Comment (jiajliu) [inline upgrading/downgrade.adoc]: When re-install v3.10, no need to install following pkgs.
atomic-openshift-master \
openvswitch \
atomic-openshift-sdn-ovs \
tuned-profiles-atomic-openshift-node

File: upgrading/downgrade.adoc (hunk 3 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: LGTM now

**Comment by jiajliu** [comment]
> @jiajliu, will you PTAL at these updates?
> 
> I made the static pod updates on this PR: #12322

Sure, I was trying to do downgrade according to #12085 and #12322. Will leave comment in pr #12322  when any issues hit for etcd restore part during downgrade.

**Comment by kalexand-rh** [comment]
Thank you @jiajliu! I fixed the package list, and you said that upgrade and downgrade part of #12322 looks good to you. Are there any more changes to downgrade, or can I merge?

**Comment by jiajliu** [comment]
> > @jiajliu, will you PTAL at these updates?
> > I made the static pod updates on this PR: #12322
> 
> Sure, I was trying to do downgrade according to #12085 and #12322. Will leave comment in pr #12322 when any issues hit for etcd restore part during downgrade.

@kalexand-rh 
About restore etcd during downgrade, I go through part "include::day_two_guide/topics/proc_restoring-etcd.adoc[leveloffset=+1]" in #12322. And leave comment in that pr.
After above steps, one of etcds config was restored, but due to atomic-openshift-node service stop before restore, so here can not just go to nest part "include::day_two_guide/topics/proc_adding-etcd-after-restoring.adoc[leveloffset=+2]". We need some actions in part "include::day_two_guide/topics/proc_bringing-openshift-online.adoc[leveloffset=+1]" to restore node service first on the etcd host, and then continue to add other members to cluster, and finally restore node services on other etcd hosts.
So you need to confirm with Dev about the correct sequences here, original three link doesn't work directly during downgrade.

File: upgrading/downgrade.adoc [hunk 8]
```diff
@@ -13,90 +13,70 @@
...
-    atomic-openshift-node \
-    atomic-openshift-master \
-    atomic-openshift-sdn-ovs \
-    atomic-openshift-excluder \
-    atomic-openshift-docker-excluder \
-    atomic-openshift-hyperkube
+   atomic-openshift-excluder \
+   atomic-openshift-hyperkube \
+   atomic-openshift-node \
+   atomic-openshift-docker-excluderh \
```
Comment (sdodson) [inline upgrading/downgrade.adoc]: typo, excluderh

File: upgrading/downgrade.adoc (hunk 5 above)
Comment (sdodson) [reply upgrading/downgrade.adoc]: I'd say that we should remove the bits about restoring etcd. Then the content of day_two_guide/topics/proc_bringing-openshift-online.adoc needs to be updated as it still references restoring control plane rpms, which is only valid for 3.10 to 3.9 downgrades.

**Comment by kalexand-rh** [comment]
@sdodson says that if you're downgrading from just 3.11 to 3.10, that you don't need to follow the etcd restore steps and that the steps to bring your cluster back online are different.

I've updated that preview to reflect those changes: http://file.rdu.redhat.com/kalexand/100918/temp_downgrade/upgrading/downgrade.html#adding-etcd-after-restoring_downgrade

Scott's PR to update the commands is here: https://github.com/openshift/openshift-docs/pull/12384

**Comment by kalexand-rh** [comment]
@jiajliu, will you PTAL?

File: upgrading/downgrade.adoc (hunk 3 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: LGTM now

File: upgrading/downgrade.adoc [hunk 9]
```diff
@@ -13,45 +13,40 @@
...
-available.
+. Locate the copies of the following files that you created when you
+xref:automated_upgrades.adoc#preparing-for-an-automated-upgrade[prepared for an upgrade].
 +
 On master hosts:
 +
 ----
-/usr/lib/systemd/system/atomic-openshift-master-api.service
-/usr/lib/systemd/system/atomic-openshift-master-controllers.service
 /etc/sysconfig/atomic-openshift-master-api
```
[the remark this answers is not in the document]
Comment (jiajliu) [reply upgrading/downgrade.adoc]: LGTM now

File: upgrading/downgrade.adoc (hunk 4 above)
Comment (jiajliu) [reply upgrading/downgrade.adoc]: LGTM now

**Comment by jiajliu** [comment]
> @jiajliu, will you PTAL?

About preview[1]：
1. If this is the latest one for downgrade, I noticed some changes were reverted, such as [Verifying backups] and [Reinstalling RPMs], compared with [2] @kalexand-rh Could u confirm?
[1] http://file.rdu.redhat.com/kalexand/100918/temp_downgrade/upgrading/downgrade.html#adding-etcd-after-restoring_downgrade
[2] https://bz1630254--ocpdocs.netlify.com/openshift-enterprise/latest/upgrading/downgrade.html
2. I've reviewed [Bringing OpenShift Container Platform services back online] section in #12384 and left comment in the pr.
3. @sdodson I noticed [Restoring etcd] section was removed from [1], so if that means etcd restore was not a must step during downgrade? So downgrade was totally separated from etcd backup/restore? If user want to restore etcd data, they can choose to follow #12322 for restore? Am I right?
> @sdodson says that if you're downgrading from just 3.11 to 3.10, that you don't need to follow the etcd restore steps and that the steps to bring your cluster back online are different.
4. @kalexand-rh Due to [Restoring etcd] section was removed, so you also need to clear etcd related in [Verifying backups].

**Comment by kalexand-rh** [comment]
@jiajliu, after some discussions with dev and @vikram-redhat, downgrade probably needs a new playbook to take care of all these issues. I'm going to merge this PR as-is and make a new PR to say that downgrade doesn't work on initial release but will be fixed with https://bugzilla.redhat.com/show_bug.cgi?id=1638004, the dev bug to create a playbook to fix downgrade.

**Comment by jiajliu** [comment]
> @jiajliu, after some discussions with dev and @vikram-redhat, downgrade probably needs a new playbook to take care of all these issues. I'm going to merge this PR as-is and make a new PR to say that downgrade doesn't work on initial release but will be fixed with https://bugzilla.redhat.com/show_bug.cgi?id=1638004, the dev bug to create a playbook to fix downgrade.

OK. Let's track left issues in bz1638004.
